pkg/shortnames/shortnames.go

```diff
@@ -0,0 +1,477 @@
+	PullCandidates []PullCandidate
+
+	userInput               string
+	unqualifiedSearchOrigin string
+	rationale               rationale
+	systemContext           *types.SystemContext
+}
+
+func (r *Resolved) addCandidate(named reference.Named) {
+	r.PullCandidates = append(r.PullCandidates, PullCandidate{named, r.userInput, false, r.systemContext})
+}
+
+func (r *Resolved) addCandidateToRecord(named reference.Named) {
+	r.PullCandidates = append(r.PullCandidates, PullCandidate{named, r.userInput, true, r.systemContext})
+}
+
+// Allows to reason over pull errors and add some context information.
+// Used in (*Resolved).WrapPullError.
+type rationale int
+
+const (
+	// No additional context.
+	rationaleNone rationale = iota
+	// Resolved value is a short-name alias.
+	rationaleAlias
+	// Resolved value has been completed with an Unqualified Search Registry.
+	rationaleUSR
+	// Resolved value has been selected by the user (via the prompt).
+	rationaleUserSelection
+)
+
+// Description returns a human-readable description about the resolution
+// process (e.g., short-name alias, unqualified-search registries, etc.).
+// It is meant to be printed before attempting to pull the pull candidates
+// to make the short-name resolution more transparent to user.
+//
+// If the returned string is empty, it is not meant to be printed.
+func (r *Resolved) Description() string {
+	switch r.rationale {
+	case rationaleAlias:
+		return fmt.Sprintf("Resolved short name %q to a recorded short-name alias (origin: %s)", r.userInput, r.unqualifiedSearchOrigin)
+	case rationaleUSR:
+		return fmt.Sprintf("Completed short name %q with unqualified-search registries (origin: %s)", r.userInput, r.unqualifiedSearchOrigin)
+	case rationaleUserSelection, rationaleNone:
+		fallthrough
+	default:
+		return ""
+	}
+}
+
+// FormatPullErrors is a convenience function to format errors that occured
+// while trying to pull one or more PullCandidates.
+func FormatPullErrors(pullErrors []error) error {
+	switch len(pullErrors) {
+	case 0:
+		return nil
+	case 1:
+		return pullErrors[0]
+	default:
+		var sb strings.Builder
+		sb.WriteString(fmt.Sprintf("%d errors occured while pulling:", len(pullErrors)))
+		for _, e := range pullErrors {
+			sb.WriteString("\n * ")
+			sb.WriteString(e.Error())
+		}
+		return errors.New(sb.String())
+	}
+}
+
+// PullCandidate is a resolved name.  Once the Value has been used
+// successfully, users MUST call `(*PullCandidate).Record(..)` to possibly
+// record it as a new short-name alias.
+type PullCandidate struct {
+	// Fully-qualified reference with tag or digest.
+	Value reference.Named
+	// Input from the user.
+	inputName string
+	// Control whether to record it permanently as an alias.
+	record bool
+	// The system context used when looking up the alias.  Always recording
+	// the system context allows for easier testing as we can simulate
+	// recording even if `record` would inititally be set false.
+	systemContext *types.SystemContext
```

Comment:
Maybe a pointer back to parent `*Resolved` instead of some of the duplication? Hopefully the GC can deal with the cycles.